cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -695,42 +689,54 @@ impl<T: Config> Pallet<T> {
 	///
 	/// On error returns a partial batch with all the XCMs processed before the failure.
 	/// This can happen in case of a decoding/re-encoding failure.
-	pub(crate) fn take_first_concatenated_xcm(
-		data: &mut &[u8],
+	pub(crate) fn take_first_concatenated_xcm<'a>(
+		data: &mut &'a [u8],
 		meter: &mut WeightMeter,
-	) -> Result<Option<BoundedVec<u8, MaxXcmpMessageLenOf<T>>>, ()> {
+	) -> Result<Option<BoundedSlice<'a, u8, MaxXcmpMessageLenOf<T>>>, ()> {
 		if data.is_empty() {
 			return Ok(None)
 		}
 
-		if meter.try_consume(T::WeightInfo::take_first_concatenated_xcm()).is_err() {
+		// Let's make sure that we can decode at least an empty xcm message.
+		if !meter.can_consume(T::WeightInfo::take_first_concatenated_xcm(0)) {
 			defensive!("Out of weight; could not decode all; dropping");
 			return Err(())
 		}
 
-		let xcm = VersionedXcm::<()>::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, data).map_err(
+		let input_data = &mut &data[..];
+		let mut input = codec::CountedInput::new(input_data);
+		VersionedXcm::<()>::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, &mut input).map_err(
 			|error| {
 				tracing::debug!(target: LOG_TARGET, ?error, "Failed to decode XCM with depth limit");
 				()
 			},
 		)?;
-		Ok(Some(xcm.encode().try_into().map_err(|error| {
-			tracing::debug!(target: LOG_TARGET, ?error, "Failed to encode XCM after decoding");
+		let (xcm_data, remaining_data) = data.split_at(input.count() as usize);
+		*data = remaining_data;
+
+		// Consume the actual weight that it took to decode this message.
+		// Saturates if it's over the limit.
+		meter.consume(T::WeightInfo::take_first_concatenated_xcm(xcm_data.len() as u32));
+
+		let xcm = Some(BoundedSlice::try_from(xcm_data).map_err(|error| {
+			tracing::debug!(target: LOG_TARGET, ?error, "Failed to take XCM after decoding");
```

Comment:
This should only happen in the case that the message is too long? Maybe promote it to an error log of even `defensive!`.

Reply:
> This should only happen in the case that the message is too long?

Yes. Added this info tho the error message.

> Maybe promote it to an error log of even defensive!.

Done. Promoted to an error log.